### PR TITLE
compiler: Introduce symbolic fencing

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

**Comment by georgebisbas** [comment]
What is the overarching target of this PR?

File: devito/ir/clusters/cluster.py [hunk 1]
```diff
@@ -212,21 +216,47 @@ def is_dense(self):
...
-                not self.is_halo_touch and
+                not self.is_wild and
                 all(a.is_regular for a in self.scope.accesses))
 
     @cached_property
     def is_sparse(self):
         """
-        A Cluster is sparse if it represents a sparse operation, i.e iff
-        There's at least one irregular access.
+        True if it represents a sparse operation, i.e iff there's at least
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/clusters/cluster.py:225]: if and only if?

File: devito/passes/clusters/misc.py [hunk 2]
```diff
@@ -44,8 +45,9 @@ def callback(self, clusters, prefix):
                 processed.append(c)
                 continue
 
-            # Synchronization operations prevent lifting
-            if c.syncs.get(dim):
+            # Synchronization prevents lifting
+            if c.syncs.get(dim) or \
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/misc.py:49]: I do the splitting on purpose for clarity when I feel there's (relatively speaking) little correlation between the two operands of the relations

**Review by EdCaunt — APPROVED** [review]

**Comment by FabioLuporini** [comment]
@georgebisbas 

> What is the overarching target of this PR?

Necessary tweaks for stuff in PRO

**Comment by FabioLuporini** [comment]
@georgebisbas 

basically it provides a series of mechanisms to impose that given subsequences of Clusters don't float around upon topological sorting, even if deemed possible by data dependence analysis

Clusters sitckiness helps with:

* performance (tough to explain though)
* code readability 
* development itself ! because it in essence creates an atomic sequence of Clusters, which is much simpler to handle in the case of asynchronous operations

I'm not sure how to explain this any better without showing the actual low-level details

**Comment by georgebisbas** [comment]
> @georgebisbas
> 
> basically it provides a series of mechanisms to impose that given subsequences of Clusters don't float around upon topological sorting, even if deemed possible by data dependence analysis
> 
> Clusters sitckiness helps with:
> 
> * performance (tough to explain though)
> * code readability
> * development itself ! because it in essence creates an atomic sequence of Clusters, which is much simpler to handle in the case of asynchronous operations
> 
> I'm not sure how to explain this any better without showing the actual low-level details

Thanks, these things could easily get into description to help review.

**Pull request merged**